library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
+        )
+
+    def install_hotfix_on_device(self):
+        params = self.want.api_params()
+        self.client.api.tm.sys.software.hotfix_s.exec_cmd(
+            'install', name=self.want.hotfix_name, **params
+        )
 
-        If the image is uploaded to the "Available Images" list but has not yet
-        been made active, this method will activate it.
+    def upload_to_device(self, filepath):
+        self.client.api.cm.autodeploy.software_image_uploads.upload_image(
+            filepath
+        )
 
-        If the image is not uploaded to the "Available Images" list, but is
-        listed as activated, this method will return true
+    def list_images_on_device(self):
+        images = self.client.api.tm.sys.software.images.get_collection()
+        return images
 
-        If the image is not uploaded to the "Available Images" list, and is not
-        in the active list, this method will upload the image and activate it.
-        """
+    def list_hotfixes_on_device(self):
+        hotfixes = self.client.api.tm.sys.software.hotfix_s.get_collection()
+        return hotfixes
 
-        changed = False
-        force = self.params['force']
-        software = self.params['software']
-        hotfix = self.params['hotfix']
-        psoftware = self.params['psoftware']
-        volume = self.params['volume']
-        reuse_inactive_volume = self.params['reuse_inactive_volume']
-
-        if reuse_inactive_volume:
-            if volume is None:
-                self.params['volume'] = self.get_next_available_volume(True)
+    def delete_image_on_device(self):
+        img = self.client.api.tm.sys.software.images.image.load(
+            name=self.want.software_name)
+        img.delete()
 
-            if self.is_activated() and not force:
-                volume = self.get_active_volume()
+    def delete_hotfix_on_device(self):
+        hf = self.client.api.tm.sys.software.hotfix_s.hotfix.load(
+            name=self.want.hotfix_name)
+        hf.delete()
 
-        if self.is_activated() and volume == self.get_active_volume():
-            return False
-        elif self.is_installed() and volume != self.get_active_volume():
-            self.api.System.SoftwareManagement.set_cluster_boot_location(volume)
-            self.api.System.Services.reboot_system(seconds_to_reboot=1)
-            self.wait_for_reboot()
-            return True
-        elif volume == self.get_active_volume():
-            raise F5ModuleError(
-                "Cannot install software or hotfixes to active volumes"
-            )
+    def list_volumes_on_device(self):
+        volumes = self.client.api.tm.sys.software.volumes.get_collection()
+        return volumes
 
-        images = self.read()
-        total = sum(len(v) for v in images.itervalues())
+    def image_exists_on_device(self):
+        images = self.client.api.tm.sys.software.images.get_collection()
+        for image in images:
```

Comment:
Since this code is repeated over and over again, i would extract it out and change this for loop into
```
if any(self._has_build_and_version(i) for i in images):
    return True
```
then define the _has_build method above as
```
if image.version != self.want.version
    return False
if image.build != self.want.build
    return False
return True
```